File: agent-editor/lib/agent_parser.py
#!/usr/bin/env python3
"""
agent_parser.py - Parse agent sessions and extract status/usage info
Configuration: Set OPENCLAW_DIR environment variable to point to your OpenClaw root.
Default: ~/.openclaw
"""

import json
import time
import os
from pathlib import Path
from datetime import datetime

# Configuration: Use OPENCLAW_DIR env var, default to ~/.openclaw
OPENCLAW_DIR = Path(os.getenv('OPENCLAW_DIR', str(Path.home() / '.openclaw')))
AGENTS_BASE = OPENCLAW_DIR / 'agents'
CONFIG_PATH = OPENCLAW_DIR / 'openclaw.json'

# Cache with 30s TTL
_cache = {
    'data': None,
    'timestamp': 0,
    'ttl': 30
}

def get_agent_list():
    """Read agent list from openclaw.json"""
    try:
        with open(CONFIG_PATH, 'r') as f:
            config = json.load(f)
        
        agents = config.get('agents', {}).get('list', [])
        default_model = config.get('agents', {}).get('defaults', {}).get('model', {}).get('primary')
        
        result = []
        for agent in agents:
            # Extract model
            model = agent.get('model')
            if isinstance(model, dict):
                model_str = model.get('primary')
            elif isinstance(model, str):
                model_str = model
            else:
                model_str = default_model
            
            # Strip provider prefix (anthropic/claude-x → claude-x)
            if model_str and '/' in model_str:
                model_str = model_str.split('/', 1)[1]
            
            result.append({
                'id': agent.get('id'),
                'model': model_str
            })
        
        return result
    except Exception as e:
        print(f"Error reading agent list: {e}")
        return []

def get_latest_session_file(agent_name):
    """Get the most recent session file for an agent"""
    sessions_dir = AGENTS_BASE / agent_name / 'sessions'
    
    if not sessions_dir.exists():
        return None
    
    try:
        # Include both active .jsonl and recently deleted ones
        all_files = list(sessions_dir.glob('*.jsonl')) + list(sessions_dir.glob('*.jsonl.deleted.*'))
        
        if not all_files:
            return None
        
        # Sort by modification time, most recent first
        all_files.sort(key=lambda f: f.stat().st_mtime, reverse=True)
        
        latest = all_files[0]
        return {
            'file': latest.name,
            'path': latest,
            'mtime': datetime.fromtimestamp(latest.stat().st_mtime),
            'deleted': '.deleted.' in latest.name
        }
    except Exception as e:
        print(f"Error finding session file for {agent_name}: {e}")
        return None

def parse_agent_session(agent_name, config_model):
    """Parse an agent's session file to extract status and usage"""
    latest_file = get_latest_session_file(agent_name)
    
    if not latest_file:
        return {
            'name': agent_name,
            'status': 'offline',
            'lastActivity': None,
            'model': config_model,
            'usage': None
        }
    
    try:
        with open(latest_file['path'], 'r') as f:
            lines = f.read().strip().split('\n')
        
        last_message = None
        last_model = None
        
        # Parse from end to find most recent message
        for line in reversed(lines):
            try:
                data = json.loads(line)
                if data.get('type') == 'message':
                    if not last_message:
                        last_message = data
                    if data.get('message', {}).get('model'):
                        last_model = data['message']['model']
                        break
            except:
                continue
        
        if not last_message:
            return {
                'name': agent_name,
                'status': 'offline',
                'lastActivity': int(latest_file['mtime'].timestamp() * 1000),
                'model': config_model,
                'usage': None
            }
        
        # Extract timestamp
        timestamp_raw = last_message.get('timestamp') or last_message.get('message', {}).get('timestamp')
        
        if timestamp_raw:
            try:
                if isinstance(timestamp_raw, (int, float)):
                    timestamp = int(timestamp_raw)
                else:
                    timestamp = int(datetime.fromisoformat(timestamp_raw.replace('Z', '+00:00')).timestamp() * 1000)
            except:
                timestamp = int(latest_file['mtime'].timestamp() * 1000)
        else:
            timestamp = int(latest_file['mtime'].timestamp() * 1000)
        
        # Determine status
        now = time.time() * 1000  # milliseconds
        age_minutes = (now - timestamp) / 60000
        
        if age_minutes < 5:
            status = 'active'
        elif age_minutes < 60:
            status = 'idle'
        else:
            status = 'offline'
        
        return {
            'name': agent_name,
            'status': status,
            'lastActivity': timestamp,
            'model': config_model,
            'usage': last_message.get('message', {}).get('usage'),
            'sessionFile': latest_file['file'],
            'fileModified': int(latest_file['mtime'].timestamp() * 1000)
        }
        
    except Exception as e:
        print(f"Error parsing session for {agent_name}: {e}")
        return {
            'name': agent_name,
            'status': 'error',
            'lastActivity': None,
            'model': config_model,
            'usage': None,
            'error': str(e)
        }

def get_agents_info():
    """Get info for all agents"""
    now = time.time()
    
    # Return cached data if still valid
    if _cache['data'] and (now - _cache['timestamp']) < _cache['ttl']:
        return _cache['data']
    
    try:
        agent_configs = get_agent_list()
        agents = [parse_agent_session(a['id'], a['model']) for a in agent_configs]
        
        # Update cache
        _cache['data'] = agents
        _cache['timestamp'] = now
        
        return agents
    except Exception as e:
        print(f"Error fetching agents info: {e}")
        return []

def get_all_session_messages():
    """Get all messages from all agent sessions"""
    agent_configs = get_agent_list()
    all_messages = []
    
    for agent_config in agent_configs:
        sessions_dir = AGENTS_BASE / agent_config['id'] / 'sessions'
        
        if not sessions_dir.exists():
            continue
        
        try:
            jsonl_files = list(sessions_dir.glob('*.jsonl'))
            
            for file_path in jsonl_files:
                try:
                    with open(file_path, 'r') as f:
                        lines = f.read().strip().split('\n')
                    
                    for line in lines:
                        try:
                            data = json.loads(line)
                            if data.get('type') == 'message':
                                timestamp_raw = data.get('timestamp') or data.get('message', {}).get('timestamp')
                                
                                if timestamp_raw:
                                    try:
                                        if isinstance(timestamp_raw, (int, float)):
                                            timestamp = int(timestamp_raw)
                                        else:
                                            timestamp = int(datetime.fromisoformat(timestamp_raw.replace('Z', '+00:00')).timestamp() * 1000)
                                    except:
                                        timestamp = None
                                else:
                                    timestamp = None
                                
                                all_messages.append({
                                    'agent': agent_config['id'],
                                    'timestamp': timestamp,
                                    'type': data.get('type'),
                                    'model': data.get('message', {}).get('model'),
                                    'usage': data.get('message', {}).get('usage'),
                                    'data': data
                                })
                        except:
                            continue
                except:
                    continue
        except:
            continue
    
    return all_messages

File: agent-editor/lib/agent-parser.js
const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);
const readFile = promisify(fs.readFile);

// Configuration: Set OPENCLAW_DIR to your OpenClaw root directory
// Default: ~/.openclaw (use process.env.OPENCLAW_DIR to override)
const OPENCLAW_DIR = process.env.OPENCLAW_DIR || path.join(process.env.HOME, '.openclaw');
const AGENTS_BASE = path.join(OPENCLAW_DIR, 'agents');
const CONFIG_PATH = path.join(OPENCLAW_DIR, 'openclaw.json');

const cache = {
    data: null,
    timestamp: 0,
    ttl: 30000 // 30 seconds
};

// Read agent list from openclaw.json
function getAgentList() {
    try {
        const config = JSON.parse(fs.readFileSync(CONFIG_PATH, 'utf-8'));
        return (config.agents?.list || []).map(a => {
            // Extract model string (handle both "string" and {"primary": "string"} formats)
            let modelStr = null;
            if (typeof a.model === 'string') {
                modelStr = a.model;
            } else if (a.model?.primary) {
                modelStr = a.model.primary;
            } else {
                modelStr = config.agents?.defaults?.model?.primary || null;
            }
            
            // Strip provider prefix if present (anthropic/claude-x → claude-x)
            if (modelStr && modelStr.includes('/')) {
                modelStr = modelStr.split('/')[1];
            }
            
            return {
                id: a.id,
                model: modelStr
            };
        });
    } catch (e) {
        return [];
    }
}

async function getLatestSessionFile(agentName) {
    const sessionsDir = path.join(AGENTS_BASE, agentName, 'sessions');
    
    try {
        const files = await readdir(sessionsDir);
        // Include both active .jsonl and recently deleted ones
        const jsonlFiles = files.filter(f => f.endsWith('.jsonl'));
        const deletedFiles = files.filter(f => f.includes('.jsonl.deleted.'));
        
        const allFiles = [...jsonlFiles, ...deletedFiles];
        if (allFiles.length === 0) return null;
        
        // Get file stats and sort by mtime
        const fileStats = await Promise.all(
            allFiles.map(async (file) => {
                const filePath = path.join(sessionsDir, file);
                try {
                    const stats = await stat(filePath);
                    return { file, path: filePath, mtime: stats.mtime, deleted: file.includes('.deleted.') };
                } catch (e) {
                    return null;
                }
            })
        );
        
        const valid = fileStats.filter(f => f !== null);
        valid.sort((a, b) => b.mtime - a.mtime);
        return valid[0] || null;
    } catch (error) {
        return null;
    }
}

async function parseAgentSession(agentName, configModel) {
    const latestFile = await getLatestSessionFile(agentName);
    
    if (!latestFile) {
        return {
            name: agentName,
            status: 'offline',
            lastActivity: null,
            model: configModel,
            usage: null
        };
    }

    try {
        const content = await readFile(latestFile.path, 'utf-8');
        const lines = content.trim().split('\n');
        
        let lastMessage = null;
        let lastModel = null;
        
        // Parse from end to find the most recent message
        for (let i = lines.length - 1; i >= 0; i--) {
            try {
                const line = JSON.parse(lines[i]);
                if (line.type === 'message') {
                    if (!lastMessage) lastMessage = line;
                    if (line.message?.model) {
                        lastModel = line.message.model;
                        break;
                    }
                }
            } catch (e) {
                continue;
            }
        }
        
        if (!lastMessage) {
            return {
                name: agentName,
                status: 'offline',
                lastActivity: latestFile.mtime.getTime(),
                model: configModel,
                usage: null
            };
        }

        const timestampRaw = lastMessage.timestamp || lastMessage.message?.timestamp;
        // Handle both ISO strings and epoch milliseconds
        let timestamp = null;
        if (timestampRaw) {
            const parsed = typeof timestampRaw === 'number' ? timestampRaw : new Date(timestampRaw).getTime();
            if (!isNaN(parsed) && parsed > 86400000) timestamp = parsed;
        }
        
        // Fallback to file mtime
        if (!timestamp) timestamp = latestFile.mtime.getTime();
        
        const now = Date.now();
        const ageMinutes = (now - timestamp) / 60000;
        
        let status = 'offline';
        if (ageMinutes < 5) status = 'active';
        else if (ageMinutes < 60) status = 'idle';
        
        return {
            name: agentName,
            status,
            lastActivity: timestamp,
            model: configModel, // Always use config model (source of truth)
            usage: lastMessage.message?.usage || null,
            sessionFile: path.basename(latestFile.path),
            fileModified: latestFile.mtime.getTime()
        };
    } catch (error) {
        console.error(`Error parsing session for ${agentName}:`, error);
        return {
            name: agentName,
            status: 'error',
            lastActivity: null,
            model: configModel,
            usage: null,
            error: error.message
        };
    }
}

async function getAgentsInfo() {
    const now = Date.now();
    
    if (cache.data && (now - cache.timestamp) < cache.ttl) {
        return cache.data;
    }

    try {
        const agentConfigs = getAgentList();
        const agents = await Promise.all(
            agentConfigs.map(a => parseAgentSession(a.id, a.model))
        );
        
        cache.data = agents;
        cache.timestamp = now;
        
        return agents;
    } catch (error) {
        console.error('Error fetching agents info:', error);
        return [];
    }
}

async function getAllSessionMessages() {
    const agentConfigs = getAgentList();
    const allMessages = [];
    
    for (const agentConfig of agentConfigs) {
        const sessionsDir = path.join(AGENTS_BASE, agentConfig.id, 'sessions');
        
        try {
            const files = await readdir(sessionsDir);
            const jsonlFiles = files.filter(f => f.endsWith('.jsonl'));
            
            for (const file of jsonlFiles) {
                const filePath = path.join(sessionsDir, file);
                try {
                    const content = await readFile(filePath, 'utf-8');
                    const lines = content.trim().split('\n');
                    
                    for (const line of lines) {
                        try {
                            const data = JSON.parse(line);
                            if (data.type === 'message') {
                                const timestampRaw = data.timestamp || data.message?.timestamp;
                                const timestamp = timestampRaw ? new Date(timestampRaw).getTime() : null;
                                
                                allMessages.push({
                                    agent: agentConfig.id,
                                    timestamp: timestamp,
                                    type: data.type,
                                    model: data.message?.model,
                                    usage: data.message?.usage,
                                    data: data
                                });
                            }
                        } catch (e) {}
                    }
                } catch (e) {}
            }
        } catch (error) {
            continue;
        }
    }
    
    return allMessages;
}

module.exports = {
    getAgentsInfo,
    getAllSessionMessages
};
